### PR TITLE
Improve the implementation of methods in PathUtils

### PR DESCRIPTION
**Pull request by bobbai00**
### What changes are proposed in this pull request?

I reimplemented the hasPrefix and findLowestCommonAncestor methods in PathUtils. According to the previous implementation, both methods will call getPathComponents methods, slice the path String into a String array and compare  each item in the String array. Such implementations will cause memory overheads and time overheads. And both methods can be implemented by directly calling String methods, including substring, equals, length. So I reimplemented both methods with String operations and comparison.

### Why are the changes needed?

hasPrefix has been called in many other methods, especially the getMountPoint of MountTable. Improve the performance of hasPrefix will also improve other methods who frequently call it.

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

File: core/common/src/test/java/alluxio/util/io/PathUtilsTest.java [hunk 1]
```diff
@@ -269,15 +291,15 @@ public void getPersistentTmpPath() {
...
+    assertArrayEquals(new String[]{"", "foo", "bar"},
         PathUtils.getPathComponents("/foo/bar"));
-    assertArrayEquals(new String[] {"", "foo", "bar"},
+    assertArrayEquals(new String[]{"", "foo", "bar"},
         PathUtils.getPathComponents("/foo/bar/"));
-    assertArrayEquals(new String[] {"", "bar"},
+    assertArrayEquals(new String[]{"", "bar"},
         PathUtils.getPathComponents("/foo/../bar"));
-    assertArrayEquals(new String[] {"", "foo", "bar", "a", "b", "c"},
+    assertArrayEquals(new String[]{"", "foo", "bar", "a", "b", "c"},
```
Comment (jiacheliu3) [inline core/common/src/test/java/alluxio/util/io/PathUtilsTest.java:302]: Restore the whitespace to avoid unnecessary changes

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 2]
```diff
@@ -283,17 +321,19 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
   public static boolean hasPrefix(String path, String prefix) throws InvalidPathException {
-    String[] pathComponents = getPathComponents(path);
-    String[] prefixComponents = getPathComponents(prefix);
-    if (pathComponents.length < prefixComponents.length) {
-      return false;
-    }
-    for (int i = 0; i < prefixComponents.length; i++) {
-      if (!pathComponents[i].equals(prefixComponents[i])) {
-        return false;
+    if (prefix.charAt(prefix.length() - 1) == '/' && path.charAt(path.length() - 1) != '/') {
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:324]: check length > 0?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 3]
```diff
@@ -283,17 +321,19 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
    * @throws InvalidPathException when the path or prefix is invalid
    */
   public static boolean hasPrefix(String path, String prefix) throws InvalidPathException {
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:323]: A short hand can be `if (prefix.equals("/") || prefix.equals("")) return true`

Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:323]: Do you need to add more unit tests on this method? How's the coverage now?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 4]
```diff
@@ -264,7 +301,8 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
     }
     if (!hasPrefix(cleanedPath, cleanedPrefix)) {
       throw new RuntimeException(
-          String.format("Cannot subtract %s from %s because it is not a prefix", prefix, path));
+          String.format("Cannot subtract %s from %s because it is not a " + "prefix",
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:304]: nit: remove unnecessary changes

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the improvements. `hasPrefix()` overall LGTM, a general comment is you can try to add a bit more comments especially when you are doing a composite if-condition check. Listing all possible cases and listing which is handled by which if-condition will help future code readers.

`findLowestCommonAncestor()` is currently used by no one. I think a string-based search may be hard, especially when you need to search for separators and be careful not to generate too many substrings. One way to totally get around that is https://github.com/Alluxio/alluxio/issues/15266 So instead of `findLowestAncestor(AlluxioURIs)`, we do `findLowestAncestor(InodePaths)` where each `InodePath` is a list of inodes with IDs. I think you can just revert the changes in `findLowestAncestor()`. I appreciate the efforts regardless :)

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 2 above)
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:298]: here I added two non-empty judgement, so that there is no need to check the length

> check length > 0?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 3 above)
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:291]: I checked the condition in `validatePath` and found that if path is "", then it is an invalid path. The condition is as follow:
```golang
boolean invalid = (path == null || path.isEmpty());
...
```
So we only need to check if `prefix` equals to "/".

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 5]
```diff
@@ -283,17 +290,27 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
       }
     }
-    return true;
+
+    if (path.length() < prefix.length()) {
+      return false;
+    }
+    return path.startsWith(prefix)
+        &&
+        (path.length() == prefix.length() || path.charAt(prefix.length()) == '/');
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:313]: Can you please add a few comments here to explain this complex logic expression? Or, you can break it into several early return if conditions, as above.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 6]
```diff
@@ -143,14 +143,21 @@ public static AlluxioURI findLowestCommonAncestor(Collection<AlluxioURI> paths)
...
       for (int i = 0; i < pathComp.length && i < matchedLen; ++i) {
         if (!matchedComponents.get(i).equals(pathComp[i])) {
+          matchedComponents = new ArrayList<>(Arrays.asList(Arrays.copyOfRange(pathComp, 0, i)));
           matchedLen = i;
           break;
         }
       }
+
+      if (matchedLen > pathComp.length) {
+        matchedComponents = new ArrayList<>(Arrays.asList(pathComp));
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:158]: same here.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 7]
```diff
@@ -143,14 +142,21 @@ public static AlluxioURI findLowestCommonAncestor(Collection<AlluxioURI> paths)
       if (matchedComponents == null) {
         matchedComponents = new ArrayList<>(Arrays.asList(pathComp));
         matchedLen = pathComp.length;
+        continue;
       }
 
       for (int i = 0; i < pathComp.length && i < matchedLen; ++i) {
         if (!matchedComponents.get(i).equals(pathComp[i])) {
+          matchedComponents = new ArrayList<>(Arrays.asList(Arrays.copyOfRange(pathComp, 0, i)));
```
[the remark this answers is not in the document]
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: You are right. There is no need to copy, `macthedLen` is sufficient to mark.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 6 above)
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: Got it. I added the comments.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 8]
```diff
@@ -283,17 +291,32 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
+    // directory(which is a `/`)
+    // can prove that `prefix` is the prefix of `path`.
+    // The second "or" condition at the second "and" condition is to exclude the
+    // case where `path` starts with the `prefix`, but the last matched position
+    // is actually not the end of a directory(e.g. prefix: '/a/b/c', path: '/a/b/ccc').
+    return path.startsWith(prefix)
+        &&
+        (path.length() == prefix.length()
+            || (prefix.endsWith("/") && path.charAt(prefix.length() - 1) == '/')
+            || path.charAt(prefix.length()) == '/');
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:319]: Your comments and the expression are getting too complex. I suggest you break each sub-expression into a boolean variable, and comment on the variable. For example:

```java
// if it starts with hello
boolean isStartsWithHello = str.startsWith("Hello,");
// and ends with world
boolean isEndsWithWorld = str.endsWith("world!");
// then it is probably "Hello, world!"
return isStartsWithHello && isEndsWithWorld;
```

Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: Got it. I seperated these conditions and comments.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 9]
```diff
@@ -283,17 +291,29 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
+    // Firstly, we check if the `path` starts with `prefix`,
+    if (path.startsWith(prefix)) {
+      // Then, check if path and prefix has the same length;
+      boolean isPathLengthEqualsPrefix = path.length() == prefix.length();
+      // This condition is to include the case like `prefix: /a/b/, path: /a/b/c/`
+      boolean isPathPrefixEnd = prefix.endsWith("/") && path.charAt(prefix.length() - 1) == '/';
+      // This condition is to exclude the case like `prefix: '/a/b/c', path: '/a/b/ccc'`
+      boolean isPathValidAfterThePrefix =
+          path.length() > prefix.length() && path.charAt(prefix.length()) == '/';
+      return isPathLengthEqualsPrefix || isPathPrefixEnd || isPathValidAfterThePrefix;
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:314]:
```suggestion
    // If prefix is "/a/" and path is "/a", make path "/a/"
    if (prefix.endsWith("/") && !path.endsWith("/")) {
      path = normalizePath(path, "/");
    }
    if (path.startsWith(prefix)) {
      return path.length() == prefix.length()  // path == prefix
          // Include cases like `prefix=/a/b/, path=/a/b/c/`
          || (prefix.endsWith("/") && path.charAt(prefix.length() - 1) == '/')
          // Exclude cases like `prefix=/a/b/c, path=/a/b/ccc`
          || (path.charAt(prefix.length()) == '/');
```

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Improved the logic a bit

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 8 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: Let's keep using chained booleans because you can get better performance by shorthand.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 9 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: Actually can you also check the length of two strings?
```
if (prefix.endsWith("/") && !path.endsWith("/") && (path.length() == prefix.length() + 1)) {
```

Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java]: The last `path.length() > prefix.length()` is not needed because it's implicitly checked already

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 3 above)
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:293]: Good catch, pls add to the comment that prefix="" is invalid

**Review by jiacheliu3 — APPROVED** [review]
LGTM with a minor comment, thanks!

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 10]
```diff
@@ -283,17 +291,26 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
    * @throws InvalidPathException when the path or prefix is invalid
    */
   public static boolean hasPrefix(String path, String prefix) throws InvalidPathException {
-    String[] pathComponents = getPathComponents(path);
-    String[] prefixComponents = getPathComponents(prefix);
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:287]: Previously, `getPathComponents` calls `cleanPath`, which "resolves dot components," i.e. `/a/b/../c` becomes `/a/c`. Your reimplementation should preserve this behaviour, or make sure paths have been cleaned up before passed to `hasPrefix`.

@jiacheliu3 watch out for this behavioural change.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 10 above)
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:287]: Yes you are right. Already removed this condition.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 11]
```diff
@@ -283,17 +291,26 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
-      return false;
+    if (prefix.isEmpty() || path.isEmpty()) {
+      throw new InvalidPathException(ExceptionMessage.PATH_INVALID.getMessage(""));
     }
-    for (int i = 0; i < prefixComponents.length; i++) {
-      if (!pathComponents[i].equals(prefixComponents[i])) {
-        return false;
-      }
+
+    if (prefix.equals("/")) {
```
[the remark this answers is not in the document]
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:298]: According to the previous implementation, it never checks the validity of the `path` and `prefix` literals. So maybe these two literals are already verified before calling hasPrefix?

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 10 above)
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:287]: Got it. I called `cleanPath` to normalize the path and prefix.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (bobbai00) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:302]: Now I used `cleanPath` to validate and normalize these two literals, so if the literal is illegal(e.g. `path` does not start with `/`), there will be an `InvalidPathException`.

[the remark this answers is not in the document]
Comment (dbw9580) [reply core/common/src/main/java/alluxio/util/io/PathUtils.java:302]: Looks good, except that catching an `﻿﻿InvalidPathException` then immediately re-throwing it does not make sense.

File: core/common/src/main/java/alluxio/util/io/PathUtils.java [hunk 12]
```diff
@@ -283,17 +291,25 @@ public static String subtractPaths(String path, String prefix) throws InvalidPat
...
+    }
+    if (!path.startsWith(prefix)) {
+      return false;
     }
-    return true;
+    return path.length() == prefix.length()  // path == prefix
+        // Include cases like `prefix=/a/b/, path=/a/b/c/`
+        || prefix.endsWith("/")
+        // Exclude cases like `prefix=/a/b/c, path=/a/b/ccc`
+        || (path.charAt(prefix.length()) == '/');
```
Comment (dbw9580) [inline core/common/src/main/java/alluxio/util/io/PathUtils.java:312]: nit: useless parentheses

**Review by dbw9580 — APPROVED** [review]
LGTM, thank you!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by dbw9580** [comment]
@jiacheliu3 cherry pick this to 2.8?

**Comment by jiacheliu3** [comment]
@Xenorith this one is a good-to-have for 2.8 as we use this utility method quite frequently. The UT coverage is quite thorough so I'm not worried.

**Comment by dbw9580** [comment]
@Xenorith candidate for cherry-picking.

**Comment by Xenorith** [comment]
i am slightly concerned with the potential impact since this is a utility method modification. although i also trust the extensive UT coverage, prefer to not cherry pick unless it is addressing a specific known problem as opposed to general nice to have. let me know if you two think otherwise

**Comment by jiacheliu3** [comment]
> i am slightly concerned with the potential impact since this is a utility method modification. although i also trust the extensive UT coverage, prefer to not cherry pick unless it is addressing a specific known problem as opposed to general nice to have. let me know if you two think otherwise

That's fine. We head start in improving 2.9.
